# DeepPokemon

In [158]:
import matplotlib.pyplot as plt
import numpy as np
import tensorflow as tf
import pandas as pd
import seaborn as sns
import plotly.graph_objects as go

In [ ]:
Import the dataset

In [3]:
file=('/Users/helenamargarida/Desktop/PokeML/Pokemon.csv')
database=pd.read_csv(file)

In [4]:
database.head()

#                   Name Type 1  Type 2  Total  HP  Attack  Defense  \
0  1              Bulbasaur  Grass  Poison    318  45      49       49   
1  2                Ivysaur  Grass  Poison    405  60      62       63   
2  3               Venusaur  Grass  Poison    525  80      82       83   
3  3  VenusaurMega Venusaur  Grass  Poison    625  80     100      123   
4  4             Charmander   Fire     NaN    309  39      52       43   

   Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0       65       65     45           1      False  
1       80       80     60           1      False  
2      100      100     80           1      False  
3      122      120     80           1      False  
4       60       50     65           1      False

Some Pokemons don't have a Type 2. We must adress these missing values.

In [10]:
database=database.fillna(value={'Type 2':'None'})

I am also going to convert the Legendary property to binary

In [12]:
from sklearn.preprocessing import LabelEncoder

labelencoder = LabelEncoder()
database['Legendary'] = labelencoder.fit_transform(database['Legendary'])

#                   Name Type 1  Type 2  Total  HP  Attack  Defense  \
0  1              Bulbasaur  Grass  Poison    318  45      49       49   
1  2                Ivysaur  Grass  Poison    405  60      62       63   
2  3               Venusaur  Grass  Poison    525  80      82       83   
3  3  VenusaurMega Venusaur  Grass  Poison    625  80     100      123   
4  4             Charmander   Fire   Blank    309  39      52       43   

   Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0       65       65     45           1          0  
1       80       80     60           1          0  
2      100      100     80           1          0  
3      122      120     80           1          0  
4       60       50     65           1          0

Let's show the Legendary Pokemon

In [87]:
Legendaries=np.array(database[database['Legendary']==1].Name)
Legendaries.size

65

We can use .describe() to explore the data

In [20]:
database.describe()

#      Total          HP      Attack     Defense     Sp. Atk  \
count  800.000000  800.00000  800.000000  800.000000  800.000000  800.000000   
mean   362.813750  435.10250   69.258750   79.001250   73.842500   72.820000   
std    208.343798  119.96304   25.534669   32.457366   31.183501   32.722294   
min      1.000000  180.00000    1.000000    5.000000    5.000000   10.000000   
25%    184.750000  330.00000   50.000000   55.000000   50.000000   49.750000   
50%    364.500000  450.00000   65.000000   75.000000   70.000000   65.000000   
75%    539.250000  515.00000   80.000000  100.000000   90.000000   95.000000   
max    721.000000  780.00000  255.000000  190.000000  230.000000  194.000000   

          Sp. Def       Speed  Generation  Legendary  
count  800.000000  800.000000   800.00000  800.00000  
mean    71.902500   68.277500     3.32375    0.08125  
std     27.828916   29.060474     1.66129    0.27339  
min     20.000000    5.000000     1.00000    0.00000  
25%     50.000000   45.000000     2.00000    0.00000  
50%     70.000000   65.000000     3.00000    0.00000  
75%     90.000000   90.000000     5.00000    0.00000  
max    230.000000  180.000000     6.00000    1.00000

So, for example, there are 800 pokemons, from generation 1 to 6. The mean Total points is 435, the weakest being 180 and the strongest 780. Which pokemons are these?

In [76]:
weakest=database.iloc[database['Total'].idxmin,:].Name

strongest=np.array(database[database['Total']==database['Total'].max()].Name)
print('The weakest pokemon overall is',weakest,'and the strongest are',strongest)

The weakest pokemon overall is Sunkern and the strongest are ['MewtwoMega Mewtwo X' 'MewtwoMega Mewtwo Y' 'RayquazaMega Rayquaza']


What about the starters? How do they compare?

In [109]:
starters=database[database['Name'].isin(['Bulbasaur','Squirtle','Charmander'])]
starters

#        Name Type 1  Type 2  Total  HP  Attack  Defense  Sp. Atk  Sp. Def  \
0  1   Bulbasaur  Grass  Poison    318  45      49       49       65       65   
4  4  Charmander   Fire   Blank    309  39      52       43       60       50   
9  7    Squirtle  Water   Blank    314  44      48       65       50       64   

   Speed  Generation  Legendary  
0     45           1          0  
4     65           1          0  
9     43           1          0

In [179]:
bulbasaur_stats=np.array(starters.iloc[0,5:11])
charmander_stats=np.array(starters.iloc[1,5:11])
squirtle_stats=np.array(starters.iloc[2,5:11])


categories = database.columns.drop(['Total','Type 1','Type 2','#','Name','Generation','Legendary'])

fig = go.Figure()

fig.add_trace(go.Scatterpolar(
      r=bulbasaur_stats,
      theta=categories,
      fill='toself',
      name='Bulbasaur',
      marker_color='#2ca02c',
      opacity=0.6
))
fig.add_trace(go.Scatterpolar(
      r=charmander_stats,
      theta=categories,
      fill='toself',
      name='Charmander',
      marker_color='tomato',
      opacity=0.6
))

fig.add_trace(go.Scatterpolar(
      r=squirtle_stats,
      theta=categories,
      fill='toself',
      name='Squirtle',
      marker_color='cornflowerblue',
      opacity=0.7
))

fig.update_layout(
  polar=dict(
    radialaxis=dict(
      visible=True,
      range=[0, 100]
    )),
  showlegend=False
)

fig.show()


Turns out Charmander is not that much greater than I thought as a kid.